---
title: "preparing gep-referenced snps dataset"
author: "Dahn-Young Dong"
date: "10/22/2020"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Visualitzation of ancestry proportions of the 16 individuals AND spatial maps of admixture coefficients

Follow the scripts from https://github.com/jdalapicolla/LanGen_pipeline_version2, the ones related to manipulating and visualizing genetic data given spatial coordinates--- 1.1 filtering SNPS and 1.2 genetic structure

All the work should be done over R and Rstudio
-----

### Visualitzation of ancestry proportions of the 16 individuals AND spatial maps of population admixture coefficients

```{r}
#load functions
source("functions_LanGen.R")
library(remotes)
library(BiocManager)
library(pacman)
library(devtools)
library(r2vcftools)
library(LEA)
library(tidyverse)
library(vcfR)
library(adegenet)
library(poppr)
library(pcadapt)
library(qvalue)
library(psych)
library(VennDiagram)

#remedy for missing package
#install.packages("devtools")
#devtools::install_github("bcm-uga/TESS3_encho_sen")
library(tess3r)

#BiocManager::install("SNPRelate")
library(dartR)

#set folders for results
create_dir(c("./adapt_var_mapping/Filtering", "./Results_Metafiles", "./Results_Filters/PCA", "./Results_Filters/PCadapt_FST", "./Results_Filters/sNMF_FST", "./Results_Filters/TESS_FST"))
```

# 1. Load Files and Verify Data Quality 

### 1.1 set up
```{r}
vcf_file = "vcf/testdata_nomissing.vcf" #biallelic SNP dataset with no missing data- from Andis Arietta 
project_name = "rasy_ymfProj"
coord_file = "coords/testdata_geoinfo.csv" ## this is using smaller dataset
coords = read.csv(coord_file)

#Define the columns positions for localities ID (local), sample ID (sample_ID), longitude, and latitude.
local = 4
sample_ID = 1
longitude = 3
latitude = 2
```

### 1.2 Load vcf file
```{r}
snps_unind <-  r2vcftools::vcfLink(vcf_file, overwriteID=T) # should I specify overwriteID as suggested in the pipeline?
```
```{r SNPs stats}
#Basic stats sample size (SZ)
capture.output(VCFsummary(snps_unind)) 
r2vcftools::Chrom(snps_unind) #return chromosome (contig), position, and ID of each SNP.
```
### 1.4 verify the quality of data

```{r}
#See Genotype Matrix
genotypes <- r2vcftools::GenotypeMatrix(snps_unind)
genotypes[1:16, 1:10] ## -1 is missing; otherwise, gives the number of derived alleles in the genotype -- ie. a 0 and 2 are both homozygotes 
```

```{r eval=FALSE, include=FALSE}
#Look at depth, quality, HWE, HE, allele frequencies, and Pi

site.depth <- r2vcftools::Query(snps_unind, type="site-mean-depth")
summary(site.depth$MEAN_DEPTH) #Mean = 33 / Median = 33
hist(site.depth$MEAN_DEPTH, breaks=200) 

quality <- Query(snps_unind, type="site-quality") 
summary(quality$QUAL) #all 13 with seemingly no variation
hist(quality$QUAL) ## no site quality information in my dataset

PI <- Query(snps_unind, type="site-pi")
mean(PI$PI) ## Mean nucleotide divergency per-site ###0.244
hist(PI$PI)

HWE <- Query(snps_unind, type="hardy")
summary(HWE$P_HWE) #Mean = 0.775 / Median = 1
hist(HWE$P_HWE)
hist(HWE$P_HWE[HWE$P_HWE<0.0001])

HE <- Query(snps_unind, type="het")
hist(HE$O.HOM) ## O.HOM, E.HOM, N_SITES, F
hist(HE$E.HOM) 
hist(HE$N_SITES) 
hist(HE$F) ## Inbreeding coefficient 0.4, quite high

freq <- Query(snps_unind, type="freq2")
hist(freq$X.FREQ.)
hist(freq$X.FREQ.1)
head(freq)
```

```{r eval=FALSE, include=FALSE}
#missing per locus
Missing <- apply(GenotypeMatrix(snps_unind), 2, function(x) sum(x < 0)/length(x)*100)
summary(Missing) ## most loci have no missing data
hist(Missing)
```

```{r eval=FALSE, include=FALSE}
#missing per individual
Missing_ind <- apply(GenotypeMatrix(snps_unind),1, function(x) sum(x<0)/length(x)*100)
summary(Missing_ind) ## Max missing = 1.4
hist(Missing_ind)
```

### 1.5 add geographic info by individuals to a vcf file

```{r}
#H. Add coordinates to vcf file.

#add the information. 
snps_unind@meta$local_ID = coords[,4] #locality pond ID
snps_unind@meta$longitude = coords[,3] #longitude
snps_unind@meta$latitude = coords[,2] #latitude
snps_unind@meta$ind_ID = coords[,1] #sample ID
#verify the file
head(snps_unind@meta) #verify
tail(snps_unind@meta) #verify
```