---
title: "structure"
author: "Dahn-Young Dong"
date: "11/24/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

_filtered_neutral.vcf is the input file, saved from the output from markdown "filter"

#------------------------------------------------------------------------------
#                       3. Genetic Structure using TESS3 
#------------------------------------------------------------------------------
# Following this tutorial: https://bcm-uga.github.io/TESS3_encho_sen/articles/main-vignette.html


```{r}
library(ggplot2)
library(raster)
library(maps)
library(rworldmap)
```

```{r}
create_dir("./Results_TESS")
```


###3.1. INPUT FILES FOR TESS:
```{r}
#A. Load the .VCF file with only neutral SNPs:
snps = vcfLink(paste0("vcf/", project_name,"_filtered_neutral.vcf"), overwriteID=T)
VCFsummary(snps) ##16 individuals and 100 SNPs.
```

```{r}
#B. Create a Genotype matrix ## repetitive from previous tutorial 
genotypes = GenotypeMatrix(snps) # only returns biallelic
genotypes[1:10, 1:10] ## -1 is missing;
class(genotypes)
genotypes = replace(genotypes, genotypes == -1, NA)
```

```{r}
#C. Create a Matrix with long and lat 
coordinates = snps@meta[,4:5]
class(coordinates)
coordinates = data.matrix(coordinates, rownames.force = NA)
class(coordinates)
#verify the coords
plot(coordinates, pch = 19, cex = .5, xlab = "Longitude", ylab = "Latitude")
```
```{r}
#F. Structure-like barplot for the Q-matrix
my.colors = c("red","orange","yellow","green","blue","purple","black","grey","turquoise") #choose the colors according to the number of K
my.palette = CreatePalette(my.colors, 9)

plot.new()
barplot(res$Q, border = NA, space = 0,
        xlab = "Individuals", ylab = "Ancestry proportions",
        main = "Ancestry matrix", col.palette = my.palette) -> bp
axis(1, at = 1:nrow(res$Q), labels = bp$order, las = 3, cex.axis = .4)
pdf("./Results_TESS/TESS_Ancestry.pdf", onefile =F)
dev.off()
```

```{r}
#G. Spatial interpolation of ancestry coefficient
pdf("./Results_TESS/TESS_MAP.pdf", onefile =F)
plot.new()
plot(res$Q, coordinates, method = "map.max", interpol = FieldsKrigModel(10),
     main = "Ancestry coefficients",cex.main=1.5,
     xlab = "Longitude", ylab = "Latitude",cex.lab=1.3,
     resolution = c(700,700), cex = 1.5,
     col.palette = my.palette)
dev.off()
```


```{r}
#H. Add admixture coefficient and replace the population ID to vcf file
Qmat_tess = as.data.frame(res$Q)
head(Qmat_tess)

columns = c() 

for (i in 1:ncol(Qmat_tess)){
  columns[i] = paste0("Adx_Coeff_TESS_", i)
}

colnames(Qmat_tess) = columns
head(Qmat_tess)
head(snps@meta)
tail(snps@meta)

for (i in 1:optimal_K){
  j = ncol(snps@meta)+1
  snps@meta[,j] = Qmat_tess[i]
}
head(snps@meta)
tail(snps@meta)

#verify individuals and maximum Adx_Coeff and add to metafile:
popIds = apply(Qmat_tess, 1, which.max)
snps@meta$PopID_tess <- popIds
head(snps@meta)
tail(snps@meta)
```
```{r}
#I. Save new vcf file with TESS results and pop ID
Save(snps, paste0("vcf/", project_name, "_filtered_neutral_LEA_DAPC_TESS.vcf")) #save vcf
write.csv(snps@meta, paste0("Results_Metafiles/ancestry_coef_LEA_DAPC_TESS_", project_name, ".csv"), quote = F) #save result as table
write.csv(as.data.frame(snps@meta$PopID_snmf), file= paste0("Results_Metafiles/", project_name, "_neutral_popsID_LEA_DAPC_TESS.csv")) #save only pop ID from sNMF

VCFsummary(snps) #16 individuals and 100 SNPs.
```


Visualize spatially interpolated ancestry coefficients together with elevation map
```{r} 
snps_neutral = vcfLink(paste0("vcf/", project_name, "_filtered_neutral_LEA_DAPC_TESS.vcf"), overwriteID=T)

par(mfrow = c(1, 2))
par(mar = c(4, 4, .1, .1))
plot(res$Q, coordinates, method = "map.max", interpol = FieldsKrigModel(10),
     main = "Ancestry coefficients",cex.main=1.5,
     xlab = "Longitude", ylab = "Latitude",cex.lab=1.3, 
     resolution = c(700,700), cex = 1.5,
     col.palette = my.palette)

#with geo and elevation (I personally stitched and wrangled the DEM file using geospatial tools during the summer)
DEM_base <- readRDS("./maps/rasters/DEM_base.rds")
plot(DEM_base, col = gray.colors(20, start = 0, end = 1), ext = ext)
points(snps_neutral@meta$longitude, snps_neutral@meta$latitude)
```

