---
title: "Dahn-Young Dong Final Project"
author: "Dahn-Young Dong"
date: "9/27/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

15% for the completed project, due by the end of reading period. The time between the presentation and submission of the final project should be used to polish remaining small issues and to incorporate feedback provided after the the presentation.

# Population Genetic Structure--Ancestry Proportions of 16 individual wood frogs each sampled from a distinct pond.

## Introduction and Goals

Yale Myer’s Forests (YMF) in northeastern Connecticut has distributions of populations of wood frogs (*Rana sylvatica*), and Skelly's lab has been recording long-term population dynamics, phenotypic and abiotic environment data that show rapid evolution is occurring (Freidenburg and Skelly, 2004; Skelly, 2004; Ligon and Skelly, 2009). That is, the populations are diverging in their physiological and morphological traits based on the local environments over generations. What requires to substantiate this ecological observation further is to explore the genetics. Previous effort to study microsatellite markers have not been successful in showing the population divergence or clustering at such microgeographical scale. 


With the availability of genome-wide SNPs for a few dozens of populations of *Rana sylvatica* from the year of 2018, and the understanding of presumably genetic-based and non-plastic local adaptations for populations that are physically in proximity to each other, few questions arise:
1. how are these unique individuals from 16 populations related to each other? 
2. qualitatively, how much correspondence is there between the individuals genetic ancestry to environmental variables, such as elevation?

My collaborator A. Z. Andis Arietta did the collection, initial filter and processing of the genomic data. And a VCF file with post-processed genomic data is available for use. The dataset contains 16 individual SNP sequences from 16 populations at focal site, YMF.

## Methods

I closely follow the scripts from https://github.com/jdalapicolla/LanGen_pipeline_version2, particularly the ones related to manipulating and visualizing genomic data given spatial coordinates--- 1.1 filtering SNPS and 1.2 genetic structure. They included 3-4 different parallel analyses and I chose TESS3 as my main analyses, which is also what the author preferred.

I used VCFtools, a linux tool, interactively linked through "r2vcftools" package in R. This is for manipulating genomic data in VCF format. I also used "TESS3R" to analyze population structure, produce ancestry proportions and spatial mapping. Lastly, I used package "sp" to create an elevation map of the locality, though the detailed wrangling is not shown in this project.  

```{r}
source("functions_LanGen.R")
library(remotes)
library(BiocManager)
library(pacman)
library(devtools)
library(r2vcftools)
library(LEA)
library(tess3r)

#set folders for results
create_dir(c("./adapt_var_mapping/Filtering", "./Results_Metafiles", "./Results_Filters/PCA", "./Results_Filters/PCadapt_FST", "./Results_Filters/sNMF_FST", "./Results_Filters/TESS_FST"))
```

### 1. Load Files and Verify Data Quality 

#### 1.1 set up
```{r}
vcf_file = "vcf/testdata_nomissing.vcf" #biallelic SNP dataset with no missing data- from Andis Arietta 
project_name = "rasy_ymfProj"
coord_file = "coords/testdata_geoinfo.csv" ## this is using smaller dataset
coords = read.csv(coord_file)
```

#### 1.2 Load vcf file
```{r}
#Creates an object of class vcfLink, which acts as an interface to a temporary VCF file. Also loads metadata (if present) associated with the samples.
snps_unind <-  r2vcftools::vcfLink(vcf_file, overwriteID=T)
```

#### 1.3 Verify the quality of data

```{r}
#Converts the VCF file to genotype matrix format and returns this matrix
genotypes <- r2vcftools::GenotypeMatrix(snps_unind)
genotypes[1:16, 1:10] ## -1 is missing; otherwise, gives the number of derived alleles in the genotype -- ie. a 0 and 2 are both homozygotes 
```

```{r eval=FALSE, include=FALSE}
#specify which analyses or conversions to perform on the data that passed through all specified filters

#Look at depth, quality, HWE, HE, allele frequencies, and Pi

site.depth <- r2vcftools::Query(snps_unind, type="site-mean-depth")
summary(site.depth$MEAN_DEPTH) #Mean = 33 / Median = 33
hist(site.depth$MEAN_DEPTH, breaks=200) 
#most SNPs have coverage of 33. This is a good number for most SNPs. 

#average number of nucleotide differences per site, denoted by pi
PI <- Query(snps_unind, type="site-pi")
mean(PI$PI) ## Mean nucleotide divergency per-site 0.244. At each SNP site, on average, 20% of the genotypes are different? Seem quite diverse. Molecular evolution at each site is quite high.
hist(PI$PI)

#Reports a p-value for each site from a Hardy-Weinberg Equilibrium test
HWE <- Query(snps_unind, type="hardy")
summary(HWE$P_HWE) #Mean = 0.775 / Median = 1
hist(HWE$P_HWE) #almost all fail to reject the null hypothesis, meaning most SNPs are at HWE, that is neutral markers under no selection
hist(HWE$P_HWE[HWE$P_HWE<0.05]) #around 800 snps our of 18000 are significantly out of HWE, violating some of the assumptions in HWE. Good sign that most SNPs are neutral here...But ideally I want to keep only loci conforming HWE
```

#### 1.5 add geographic info by individuals to a vcf file

```{r}
#Add coordinates to vcf file.

#add the information. 
snps_unind@meta$local_ID = coords[,4] #locality pond ID
snps_unind@meta$longitude = coords[,3] #longitude
snps_unind@meta$latitude = coords[,2] #latitude
snps_unind@meta$ind_ID = coords[,1] #sample ID
#verify the file
head(snps_unind@meta) #verify
```


### 2. Load Files and Verify Data Quality 

#### 2.1 filter dataset by missingness, min and max coverage, and hardy-weinberg equilibrium (HWE):
```{r}
#From dataset with all individuals
snps_fil <- Filter(snps_unind, filterOptions(max.missing = 1, min.meanDP=20, max.meanDP=500, hwe=0.7)) #max.missing---Exclude sites on the basis of the proportion of missing data (defined to be between 0 and 1, where 0 allows sites that are completely missing and 1 indicates no missing data allowed) #chose hwe=0.7 to rule out all possible non-HWE loci, based on the quality check above. 

VCFsummary(snps_fil)  ## 16 individuals and 9163 SNPs left.
```

#### 2.2 filter dataset by linkage disequilibrium (LD) between contigs.

```{r}
#Define R² value:
r2 = 0.7 # filtering out small degree of linked loci but still maximizing available SNPs left for analyses
#remove snps with R²
#ld_between <- Linkage(snps_fil, type="interchrom-geno-r2", linkageOptions(min.r2=r2))
ld_between <- read.csv(paste0("adapt_var_mapping/Filtering/ld_between_", r2, "_hwe_test2.csv")) #load this file if it was saved before
head(ld_between)
hist(ld_between$R.2)
#write.csv(ld_between, file= paste0("adapt_var_mapping/Filtering/ld_between_", r2, "_hwe_test2.csv"))
```

```{r}
#Select one set of the correlated snps (ID1 or ID2) # I went for ID2
ld2_snps <- ld_between$ID2
nold2_snps <- snps_fil@site_id[!(snps_fil@site_id %in% ld2_snps)]
snps_fil_ldF <- Subset(snps_fil, sites=nold2_snps) # Keep snps that are not in LD.
neutralLDBetween_SZ = capture.output(VCFsummary(snps_fil_ldF)) 
neutralLDBetween_SZ # "16 individuals and 2796 SNPs."
```

#### 2.3 verify the quality of the filtered dataset

```{r}
site.depth2 <- Query(snps_fil_ldF, "site-mean-depth")
HWE2 <- Query(snps_fil_ldF, type="hardy")

hist(site.depth2$MEAN_DEPTH, breaks=50)
hist(site.depth2$MEAN_DEPTH[site.depth2$MEAN_DEPTH <200])
hist(HWE2$P_HWE[HWE2$P_HWE<200])
```


## Results

The tree in Figure 1...

## Discussion

These results indicate...

The biggest difficulty in implementing these analyses was...

If I did these analyses again, I would...

## References

Freidenburg, L. K., and Skelly, D. K. (2004). Microgeographical variation in thermal preference by an amphibian. Ecol. Lett. 7, 369–373. doi:10.1111/j.1461-0248.2004.00587.x.

Ligon, N. F., and Skelly, D. K. (2009). Cryptic divergence: countergradient variation in the wood frog. Evol. Ecol. Res. 11, 1099–1109. http://www.evolutionary-ecology.com/issues/v11/n07/jjar2510.pdf

Skelly, D. K. (2004). Microgeographic countergradient variation in the wood frog, Rana sylvatica. Evolution 58, 160–165. doi:10.1111/j.0014-3820.2004.tb01582.x.

The Variant Call Format and VCFtools, Petr Danecek, Adam Auton, Goncalo Abecasis, Cornelis A. Albers, Eric Banks, Mark A. DePristo, Robert Handsaker, Gerton Lunter, Gabor Marth, Stephen T. Sherry, Gilean McVean, Richard Durbin and 1000 Genomes Project Analysis Group, Bioinformatics, 2011

Kevin Caye, Timo Deist, Helena Martins, Olivier Michel, Olivier Francois (2016) TESS3: fast inference of spatial population structure and genome scans for selection. Molecular Ecology Resources 16 (2), 540-548. doi: 10.1111/1755-0998.12471.

Kevin Caye, Flora Jay, Olivier Michel, Olivier Francois. Fast Inference of Individual Admixture Coefficients Using Geographic Data. bioRxiv, 2016, doi: http://dx.doi.org/10.1101/080291

Helena Martins, Kevin Caye, Keurcien Luu, Michael GB Blum, Olivier Francois (2016). Identifying outlier loci in admixed and in continuous populations using ancestral population differentiation statistics. Molecular Ecology 25 (20), 5029-5042. doi: http://dx.doi.org/10.1101/054585.
