---
title: "Dahn-Young Dong Final Project"
author: "Dahn-Young Dong"
date: "9/27/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Each student will conduct their own phylogenetic study, which could include development of a new method, reconstructing the phylogeny of a particular group of organisms, examining the evolution of one more more characters on a phylogeny, or examination of the behavior of a particular phylogenetic method. The professor and other students will review the project as it is developed so that it can be iteratively improved, tested, and refined. At the end of the course, each student will record a video (about 10 minutes long) about the project to share with the class, and in the last days of class each student will lead a discussion of their project for 5-10 minutes. The points for the final project will not be awarded all at once, but cumulatively over the semester:

5% in Week 6 for the project proposal. The project proposal consists of filling out the "Introduction and Goals" section, which specifies the question to be addressed, the methods to be applied, and the data (if any) to be considered.
5% in Week 9 for the project minimum viable analysis, a worked example that shows how all the concepts and tools will be connected.
10% for the recorded presentation to be shown in the last class. The recording of the presentation is due right after November recess. It is expected that this presentation covers the nearly complete project and that all major analyses are done. Presentation grade will reflect both the delivery of the material, as well as the state and quality of the work in the repository.
15% for the completed project, due by the end of reading period. The time between the presentation and submission of the final project should be used to polish remaining small issues and to incorporate feedback provided after the the presentation.

# Population tree construction and spatially-explicit admixture mapping

## Introduction and Goals

Yale Myer’s Forests (YMF) in northeastern Connecticut has distributions of populations of wood frogs (*Rana sylvatica*), and Skelly's lab has been recording long-term population dynamics, phenotypic and abiotic environment data that show rapid evolution is occurring (Freidenburg and Skelly, 2004; Skelly, 2004; Ligon and Skelly, 2009). That is, the populations are diverging in their physiological and morphological traits based on the local environments over generations. What requires to substantiate this ecological observation further is to explore the genetics. Previous effort to study microsatellite markers have not been successful in showing the population divergence or clustering at such microgeographical scale. 


With the availability of genome-wide SNPs for a few dozens of populations of *Rana sylvatica* from the year of 2018, and the understanding of presumably genetic-based and non-plastic local adaptations for populations that are physically in proximity to each other, few questions arise:
1. how are these populations related to each other phylogenetically, 
2. which groups are older lineages and which groups are later colonized (admixture),
3. what is the spatial pattern of current population distribution and colonial history, and is there a clear narrative?

My collaborator A. Z. Andis Arietta did the collection, filter and processing of the genomic data. And a VCF file with post-processed genomic data is available for use. The dataset contains 277 individual SNP sequences from 41 populations at focal site, YMF, and then three outgroup populations from New Haven Co. and Guilford Co. Connecticut. The are a few things and related methods I will need to approach my questions.

1. Access and view the dataset (Rstudio, High-Computing Cluster Platform)
2. Population tree-building and visualization (IQtree), which involves naming convention changes, evolution model selection and parameterization, tree model evaluation, and annotation
3. Spatial mapping of population with phylogenetic and genotype proportion information (using package "gstudio")
4. Genetic neighbor analysis and admixture analysis

Completion of 1 and 2 would be satisfactory for my final project, inclusion of successful anaslysis for 3 would be well done. And possible attempt for 4 would be extraordinary.

## Methods

The tools I used were... See analysis files at (links to analysis files).

## Results

The tree in Figure 1...

## Discussion

These results indicate...

The biggest difficulty in implementing these analyses was...

If I did these analyses again, I would...

## References

Freidenburg, L. K., and Skelly, D. K. (2004). Microgeographical variation in thermal preference by an amphibian. Ecol. Lett. 7, 369–373. doi:10.1111/j.1461-0248.2004.00587.x.

Ligon, N. F., and Skelly, D. K. (2009). Cryptic divergence: countergradient variation in the wood frog. Evol. Ecol. Res. 11, 1099–1109.

Skelly, D. K. (2004). Microgeographic countergradient variation in the wood frog, Rana sylvatica. Evolution 58, 160–165. doi:10.1111/j.0014-3820.2004.tb01582.x.
