---
title: "Minimum viable analysis"
author: "Dahn-Young Dong"
date: "10/22/2020"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Mimimum Viable Analysis

I: Use program TreeMix to make inference of patterns of population splitting and mixing from genomewide allele frequency data (over Linux Shell bash either locally or HPC grace to understand gene flow directions of populations)

Source for the documentation and download of treemix https://bitbucket.org/nygcresearch/treemix/downloads/
script example by github contributor https://github.com/jbv2/nf-vcf2treemix

II: Spatially visualizing genetic clusters with maps of population admixture coefficients

Follow the scripts from https://github.com/jdalapicolla/LanGen_pipeline_version2, the ones related to manipulating and visualizing genetic data given spatial coordinates--- 1.1 filtering SNPS, 1.2 genetic structure, 1.3 Maps.

All the work should be done over R and Rstudio
-----

### Spatially visualizing genetic clusters with maps of population admixture coefficients

```{r}
#load functions
source("functions_LanGen.R")

if("remotes" %in% rownames(installed.packages()) == FALSE){install.packages("remotes")
} else {print (paste0("'remotes' has already been installed in library"))}
if("BiocManager" %in% rownames(installed.packages()) == FALSE){install.packages("BiocManager")
} else {print (paste0("'BiocManager' has already been installed in library"))}
if("pacman" %in% rownames(installed.packages()) == FALSE){install.packages("pacman")
} else {print (paste0("'pacman' has already been installed in library"))}
if("devtools" %in% rownames(installed.packages()) == FALSE){install.packages("devtools")
} else {print (paste0("'devtools' has already been installed in library"))}

if("r2vcftools" %in% rownames(installed.packages()) == FALSE){remotes::install_github("nspope/r2vcftools")
} else {print (paste0("'r2vcftools' has already been installed in library"))}
if("LEA" %in% rownames(installed.packages()) == FALSE){BiocManager::install("LEA")
} else {print (paste0("'LEA' has already been installed in library"))}
if("tidyverse" %in% rownames(installed.packages()) == FALSE){install.packages("tidyverse")
} else {print (paste0("'tidyverse' has already been installed in library"))}
if("vcfR" %in% rownames(installed.packages()) == FALSE){install.packages("vcfR")
} else {print (paste0("'vcfR' has already been installed in library"))}
if("dartR" %in% rownames(installed.packages()) == FALSE){install.packages("dartR")
} else {print (paste0("'dartR' has already been installed in library"))}
if("adegenet" %in% rownames(installed.packages()) == FALSE){install.packages("adegenet")
} else {print (paste0("'adegenet' has already been installed in library"))}
if("poppr" %in% rownames(installed.packages()) == FALSE){install.packages("poppr")
} else {print (paste0("'poppr' has already been installed in library"))}
if("pcadapt" %in% rownames(installed.packages()) == FALSE){install.packages("pcadapt")
} else {print (paste0("'pcadapt' has already been installed in library"))}
if("qvalue" %in% rownames(installed.packages()) == FALSE){BiocManager::install("qvalue")
} else {print (paste0("'qvalue' has already been installed in library"))}
if("psych" %in% rownames(installed.packages()) == FALSE){install.packages("psych")
} else {print (paste0("'psych' has already been installed in library"))}
if("tess3r" %in% rownames(installed.packages()) == FALSE){devtools::install_github("bcm-uga/TESS3_encho_sen")
} else {print (paste0("'tess3r' has already been installed in library"))}
if("VennDiagram" %in% rownames(installed.packages()) == FALSE){install.packages("VennDiagram")
} else {print (paste0("'VennDiagram' has already been installed in library"))}

#remedy for missing package
pacman::p_load(tess3r)
#r2vcftools is not available in 3.6.2 r version
library(r2vcftools)

#set folders for results
create_dir(c("./adapt_var_mapping/Filtering", "./Results_Metafiles", "./Results_Filters/PCA", "./Results_Filters/PCadapt_FST", "./Results_Filters/sNMF_FST", "./Results_Filters/TESS_FST"))
```

# 1. Load Files and Verify Data Quality 

### 1.1 set up
```{r}
vcf_file = "vcf/testdata.vcf" #this departs from the flow from Nate about genind object-
project_name = "rasy_ymfProj"
coord_file = "coords/testdata_geoinfo.csv" ## this is using smaller dataset, ramdom matching of pond and trimnames!!!!
coords = read.csv(coord_file)

#Define the columns positions for localities ID (local), sample ID (sample_ID), longitude, and latitude.
local = 4
sample_ID = 1
longitude = 3
latitude = 2
```

### 1.2 Load vcf file
```{r}
snps_raw <-  r2vcftools::vcfLink(vcf_file, overwriteID=T) # should I specify overwriteID as suggested in the pipeline?
```

```{r}
#Remove loci that are not SNPs
snps <- Filter(snps_raw, filterOptions(max.alleles=2, min.alleles=2), indels="remove")
```

```{r SNPs stats}
#Basic stats sample size (SZ)
raw_SZ = capture.output(VCFsummary(snps_raw)) 
filtered_SZ = capture.output(VCFsummary(snps)) 
raw_SZ #16 individuals and 227501 SNPs. 
filtered_SZ #16 individuals and 223601 SNPs.
snps@sample_id
snps@site_id
snps@meta
r2vcftools::Chrom(snps) #return chromosome (contig), position, and ID of each SNP.
```


### 1.3 Remove individuals IF IT IS NECESSARY
```{r eval=FALSE, include=FALSE}
#There are three duplicates in snps@sample_id,

#Remove one individual
snps@meta$sample_name
UNIND <- snps@sample_id[snps@sample_id != "9202_rep_sorted"]
snps_unind <- Subset(snps, samples=UNIND)
snps_unind@meta
VCFsummary(snps_unind) ## 290 individuals and 35214 SNPs.

#B. Remove more than one individual. In the case Granito locality, the rep individual (9202_rep), and individuals (9541, 9561):
snps@meta$sample_name
remove_ind = c("9202_rep_sorted", "9541_sorted", "9561_sorted", "9521_sorted","9522_sorted", "9523_sorted", "9526_sorted","9530_sorted","9531_sorted", "9532_sorted","9533_sorted","9537_sorted","9538_sorted","9540_sorted")
UNIND <- snps@sample_id[!snps@sample_id %in% remove_ind]
snps_unind <- Subset(snps, samples=UNIND)

#verify sample size
length(snps@meta$sample_name)-length(remove_ind)
length(snps_unind@meta$sample_name)
VCFsummary(snps_unind) ## 277 individuals and 35214 SNPs.
```

In my case, let's assume I don't remove any individuals from the vcf file.
```{r}
# If you do not need to remove individuals,
snps_unind = snps
VCFsummary(snps_unind) ## 16 individuals and 223601 SNPs.
```

### 1.4 verify the quality of data

If the saved snps_unind object is successful, I can resume anytime from here.
```{r eval=FALSE, include=FALSE}
#access the output from working directory
#saveRDS(snps_unind, "./output/snps_unind.rds")
snps_unind <- readRDS("./output/snps_unind.rds")
```

```{r}
#See Genotype Matrix
genotypes <- r2vcftools::GenotypeMatrix(snps_unind)
genotypes[1:10, 1:10] ## -1 is missing; otherwise, gives the number of derived alleles in the genotype -- ie. a 0 and 2 are both homozygotes 
#saveRDS(genotypes, "./output/genotypes.rds")
#genotypes <- readRDS("./output/genotypes.rds")
```

```{r}
#Look at depth, quality, HWE, HE, allele frequencies, and Pi

site.depth <- r2vcftools::Query(snps_unind, type="site-mean-depth")
summary(site.depth$MEAN_DEPTH) #Mean = 17 / Median = 12
hist(site.depth$MEAN_DEPTH, breaks=200) 

quality <- Query(snps_unind, type="site-quality") 
summary(quality$QUAL) #all 13 with seemingly no variation
hist(quality$QUAL) ## something wrong here.
#---------------------------------- analysed till here Nov 17th, 2020
PI <- Query(snps_unind, type="site-pi")
mean(PI$PI) ## Mean nucleotide divergency per-site ###0.29
hist(PI$PI)

HWE <- Query(snps_unind, type="hardy")
summary(HWE$P_HWE) #Mean = 0.129 / Median = 0.0012
hist(HWE$P_HWE)
hist(HWE$P_HWE[HWE$P_HWE<0.0001])

HE <- Query(snps_unind, type="het")
hist(HE$O.HOM) ## O.HOM, E.HOM, N_SITES, F
hist(HE$E.HOM) 
hist(HE$N_SITES) 
hist(HE$F) ## Inbreeding coefficient 0.4, quite high

freq <- Query(snps_unind, type="freq2")
hist(freq$X.FREQ.)
hist(freq$X.FREQ.1)
head(freq)
```

```{r eval=FALSE, include=FALSE}
#missing per locus
Missing <- apply(GenotypeMatrix(snps_unind), 2, function(x) sum(x < 0)/length(x)*100)
summary(Missing) ## no missingness??
hist(Missing)
```

```{r eval=FALSE, include=FALSE}
#missing per individual
Missing_ind <- apply(GenotypeMatrix(snps_unind),1, function(x) sum(x<0)/length(x)*100)
summary(Missing_ind) ## Max missing = 0
hist(Missing_ind)
```
### 1.5 add geographic info and missing data by individuals to a vcf file

```{r eval=FALSE, include=FALSE}
#Save missind data information as data frame # I have no missing data though.
library(stringr)
Missingind.df = as.data.frame(Missing_ind)
#remove string "_sorted" from samples ID if you need it
row.names(Missingind.df) = str_remove(row.names(Missingind.df), pattern="_sorted")
#add samples ID to the data frame
Missingind.df$ID = row.names(Missingind.df)
Missingind.df
```

```{r eval=FALSE, include=FALSE}
#Remove geographical information of individuals that you excluded in step #3 
#no need because I didn't exclude any
missing_coord = coords[coords[,sample_ID] %in% Missingind.df$ID,]
head(missing_coord)
tail(missing_coord)
length(missing_coord[,1])
```

#####When coords are in UTM you need to transform to lon/lat in decimals. Use this code:
```{r}
missing_coord = missing_coord[!is.na(missing_coord[,longitude]),]
```

```{r eval=FALSE, include=FALSE}
#Add a datum
#My UTM data are in SAD69 UTM zone 22S. Search in internet the EPSG code for that area. For example https://epsg.io/ 
#In the case is EPSG:32722.
#library(aqp)
#coord_gen_pts = missing_coord[,c(longitude, latitude)] #select lat and long
#coordinates(coord_gen_pts) <- coord_gen_pts
#projection(coord_gen_pts) = crs("+init=epsg:32722") #define utm datum only for coord cols
```

```{r}
#Sort the geographical information file according to the sample order of the missing data file:
missing_coord_sorted = missing_coord[order(match(missing_coord[,sample_ID], Missingind.df$ID)),]
head(missing_coord_sorted)
```

```{r}
#Check if there are some difference between files. If they are corrected, function will return "character (0)"
setdiff(Missingind.df$ID, missing_coord_sorted[,sample_ID])
```

```{r}
#Check if identical samples were selected in the same order. If they are corrected, function will return "TRUE"
identical(as.character(Missingind.df$ID), as.character(missing_coord_sorted[,sample_ID]))
```

```{r}
#Create a data frame merging missing and geographical information
missing_local = cbind(missing_coord_sorted[,c(local, longitude, latitude)], Missingind.df)
head(missing_local)
tail(missing_local)
```














### TreeMix