---
title: "Minimum viable analysis"
author: "Dahn-Young Dong"
date: "10/22/2020"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Mimimum Viable Analysis

I: Use program TreeMix to make inference of patterns of population splitting and mixing from genomewide allele frequency data (over Linux Shell bash either locally or HPC grace to understand gene flow directions of populations)

Source for the documentation and download of treemix https://bitbucket.org/nygcresearch/treemix/downloads/
script example by github contributor https://github.com/jbv2/nf-vcf2treemix

II: Spatially visualizing genetic clusters with maps of population admixture coefficients

Follow the scripts from https://github.com/jdalapicolla/LanGen_pipeline_version2, the ones related to manipulating and visualizing genetic data given spatial coordinates--- 1.1 filtering SNPS, 1.2 genetic structure, 1.3 Maps.

All the work should be done over R and Rstudio
-----
Updates for Instructors: 

For I: 
I got TREEMIX to run over cluster

For II: I am figuring out how to use regular expression to match the sample names in order to run the pipeline. 

### Spatially visualizing genetic clusters with maps of population admixture coefficients

```{r}
#load package
library(pegas)
library(adegenet)
library(vcfR)
library(devtools)
library(ade4)
library(spdep)
library(vegan)
library(dplyr)
library(MASS)
library(sf)
```

```{r}
INDDATA <- "coords/YMF2018_RASYpopgen_sampleinfo.csv"
#frogGenind <- load("./large.data/frogGenind.Rdata") # if I retrieve frogGenind in this manner, the indNames function below won't work. I had to rerun the vcfR2genind function which takes a long while.

#VCFFILE <- "vcf/rasy_ymf_main.vcf"
#frogVariants <- vcfR::read.vcfR(VCFFILE, checkFile = T,verbose = T)
#frogGenind <- vcfR2genind(frogVariants)

# Cleaning data (code borrowed from Nate Edelman)
getGoodID <- function(x){
  parts=unlist(strsplit(x,"_"))
  goodParts=parts[parts!="YNL"]
  finalName=head(goodParts,-1)
  return(paste(finalName[1],finalName[2],sep="_"))
}

sampleData <- read.csv(INDDATA) %>%
  mutate(sampleName=paste(Pop,sprintf("%03d",Extract),sep="_"))
sampleData <- sampleData[2:ncol(sampleData)] %>%
  distinct(.)
distances <- sampleData[,c("Pond","Lat","Long")] %>% 
  distinct(.) 
distances <- distances[order(distances$Pond),]

validIDs=(as.vector(sapply(indNames(frogGenind),getGoodID))) ## indNames require proper loading of frogGenind object. 
sampleInObject <- sampleData[sampleData$sampleName %in% validIDs,]
```

```{r}
# set basic geospatial reference data
geoinfo <- sampleInObject[,c(14,12,13,5)] 
names(geoinfo)[c(2,3)]<- c("latitude","longitude")
#IF TWO OR MORE SAMPLES WERE COLLECTED IN THE SAME LOCALITY, CHANGE THE VALUE OF 4TH OR 5TH DECIMAL NUMBER TO SET A DIFFERENCE OF 1 OR 5 METERS. SOME ANALYSES COULD RETURN ERRORS IF THE DIFFERENCES IN METERS BETWEEN INDIVIDUALS IS 0.

#load functions
source("functions_LanGen.R")

if("remotes" %in% rownames(installed.packages()) == FALSE){install.packages("remotes")
} else {print (paste0("'remotes' has already been installed in library"))}
if("BiocManager" %in% rownames(installed.packages()) == FALSE){install.packages("BiocManager")
} else {print (paste0("'BiocManager' has already been installed in library"))}
if("pacman" %in% rownames(installed.packages()) == FALSE){install.packages("pacman")
} else {print (paste0("'pacman' has already been installed in library"))}
if("devtools" %in% rownames(installed.packages()) == FALSE){install.packages("devtools")
} else {print (paste0("'devtools' has already been installed in library"))}

if("r2vcftools" %in% rownames(installed.packages()) == FALSE){remotes::install_github("nspope/r2vcftools")
} else {print (paste0("'r2vcftools' has already been installed in library"))}
if("LEA" %in% rownames(installed.packages()) == FALSE){BiocManager::install("LEA")
} else {print (paste0("'LEA' has already been installed in library"))}
if("tidyverse" %in% rownames(installed.packages()) == FALSE){install.packages("tidyverse")
} else {print (paste0("'tidyverse' has already been installed in library"))}
if("vcfR" %in% rownames(installed.packages()) == FALSE){install.packages("vcfR")
} else {print (paste0("'vcfR' has already been installed in library"))}
if("dartR" %in% rownames(installed.packages()) == FALSE){install.packages("dartR")
} else {print (paste0("'dartR' has already been installed in library"))}
if("adegenet" %in% rownames(installed.packages()) == FALSE){install.packages("adegenet")
} else {print (paste0("'adegenet' has already been installed in library"))}
if("poppr" %in% rownames(installed.packages()) == FALSE){install.packages("poppr")
} else {print (paste0("'poppr' has already been installed in library"))}
if("pcadapt" %in% rownames(installed.packages()) == FALSE){install.packages("pcadapt")
} else {print (paste0("'pcadapt' has already been installed in library"))}
if("qvalue" %in% rownames(installed.packages()) == FALSE){BiocManager::install("qvalue")
} else {print (paste0("'qvalue' has already been installed in library"))}
if("psych" %in% rownames(installed.packages()) == FALSE){install.packages("psych")
} else {print (paste0("'psych' has already been installed in library"))}
if("tess3r" %in% rownames(installed.packages()) == FALSE){devtools::install_github("bcm-uga/TESS3_encho_sen")
} else {print (paste0("'tess3r' has already been installed in library"))}
if("VennDiagram" %in% rownames(installed.packages()) == FALSE){install.packages("VennDiagram")
} else {print (paste0("'VennDiagram' has already been installed in library"))}

#remedy for missing package
pacman::p_load(tess3r)
#r2vcftools is not available in 3.6.2 r version

#set folders for results
create_dir(c("./adapt_var_mapping/Filtering", "./Results_Metafiles", "./Results_Filters/PCA", "./Results_Filters/PCadapt_FST", "./Results_Filters/sNMF_FST", "./Results_Filters/TESS_FST"))
```

# 1. Load Files and Verify Data Quality 
```{r}
vcf_file = "vcf/rasy_ymf_main.vcf" #this departs from the flow from Nate about genind object-
project_name = "rasy_ymfProj"
coord_file = "coords/geoinfo.csv"
coords = read.csv(coord_file)

#Define the columns positions for localities ID (local), sample ID (sample_ID), longitude, and latitude.
local = 5
sample_ID = 2
longitude = 4
latitude = 3
```

```{r}
library(r2vcftools)
#Load vcf file
snps_raw <-  r2vcftools::vcfLink(vcf_file, overwriteID=T) # should I specify overwriteID as suggested in the pipeline?
snps_raw
```

```{r}
#Remove loci that are not SNPs
snps <- Filter(snps_raw, filterOptions(max.alleles=2, min.alleles=2), indels="remove")
```

```{r}

```

### TreeMix