---
title: "Minimum viable analysis"
author: "Dahn-Young Dong"
date: "10/22/2020"
output: rmarkdown::github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Mimimum Viable Analysis

I: Use program TreeMix to make inference of patterns of population splitting and mixing from genomewide allele frequency data (over Linux Shell bash either locally or HPC grace to understand gene flow directions of populations)

Source for the documentation and download of treemix https://bitbucket.org/nygcresearch/treemix/downloads/
script example by github contributor https://github.com/jbv2/nf-vcf2treemix

II: Spatially visualizing genetic clusters with maps of population admixture coefficients

Follow the scripts from https://github.com/jdalapicolla/LanGen_pipeline_version2, the ones related to manipulating and visualizing genetic data given spatial coordinates--- 1.1 filtering SNPS, 1.2 genetic structure, 1.3 Maps.

All the work should be done over R and Rstudio
-----

### Spatially visualizing genetic clusters with maps of population admixture coefficients

```{r}
#load functions
source("functions_LanGen.R")

if("remotes" %in% rownames(installed.packages()) == FALSE){install.packages("remotes")
} else {print (paste0("'remotes' has already been installed in library"))}
if("BiocManager" %in% rownames(installed.packages()) == FALSE){install.packages("BiocManager")
} else {print (paste0("'BiocManager' has already been installed in library"))}
if("pacman" %in% rownames(installed.packages()) == FALSE){install.packages("pacman")
} else {print (paste0("'pacman' has already been installed in library"))}
if("devtools" %in% rownames(installed.packages()) == FALSE){install.packages("devtools")
} else {print (paste0("'devtools' has already been installed in library"))}

if("r2vcftools" %in% rownames(installed.packages()) == FALSE){remotes::install_github("nspope/r2vcftools")
} else {print (paste0("'r2vcftools' has already been installed in library"))}
if("LEA" %in% rownames(installed.packages()) == FALSE){BiocManager::install("LEA")
} else {print (paste0("'LEA' has already been installed in library"))}
if("tidyverse" %in% rownames(installed.packages()) == FALSE){install.packages("tidyverse")
} else {print (paste0("'tidyverse' has already been installed in library"))}
if("vcfR" %in% rownames(installed.packages()) == FALSE){install.packages("vcfR")
} else {print (paste0("'vcfR' has already been installed in library"))}
if("dartR" %in% rownames(installed.packages()) == FALSE){install.packages("dartR")
} else {print (paste0("'dartR' has already been installed in library"))}
if("adegenet" %in% rownames(installed.packages()) == FALSE){install.packages("adegenet")
} else {print (paste0("'adegenet' has already been installed in library"))}
if("poppr" %in% rownames(installed.packages()) == FALSE){install.packages("poppr")
} else {print (paste0("'poppr' has already been installed in library"))}
if("pcadapt" %in% rownames(installed.packages()) == FALSE){install.packages("pcadapt")
} else {print (paste0("'pcadapt' has already been installed in library"))}
if("qvalue" %in% rownames(installed.packages()) == FALSE){BiocManager::install("qvalue")
} else {print (paste0("'qvalue' has already been installed in library"))}
if("psych" %in% rownames(installed.packages()) == FALSE){install.packages("psych")
} else {print (paste0("'psych' has already been installed in library"))}
if("tess3r" %in% rownames(installed.packages()) == FALSE){devtools::install_github("bcm-uga/TESS3_encho_sen")
} else {print (paste0("'tess3r' has already been installed in library"))}
if("VennDiagram" %in% rownames(installed.packages()) == FALSE){install.packages("VennDiagram")
} else {print (paste0("'VennDiagram' has already been installed in library"))}

#remedy for missing package
install.packages("devtools")
devtools::install_github("bcm-uga/TESS3_encho_sen")
library(r2vcftools)

#set folders for results
create_dir(c("./adapt_var_mapping/Filtering", "./Results_Metafiles", "./Results_Filters/PCA", "./Results_Filters/PCadapt_FST", "./Results_Filters/sNMF_FST", "./Results_Filters/TESS_FST"))
```

# 1. Load Files and Verify Data Quality 

### 1.1 set up
```{r}
vcf_file = "vcf/testdata.vcf" #this departs from the flow from Nate about genind object-
project_name = "rasy_ymfProj"
coord_file = "coords/testdata_geoinfo.csv" ## this is using smaller dataset, ramdom matching of pond and trimnames!!!!
coords = read.csv(coord_file)

#Define the columns positions for localities ID (local), sample ID (sample_ID), longitude, and latitude.
local = 4
sample_ID = 1
longitude = 3
latitude = 2
```

### 1.2 Load vcf file
```{r}
snps_raw <-  r2vcftools::vcfLink(vcf_file, overwriteID=T) # should I specify overwriteID as suggested in the pipeline?
```

```{r}
#Remove loci that are not SNPs
snps <- Filter(snps_raw, filterOptions(max.alleles=2, min.alleles=2), indels="remove")
```

```{r SNPs stats}
#Basic stats sample size (SZ)
raw_SZ = capture.output(VCFsummary(snps_raw)) 
filtered_SZ = capture.output(VCFsummary(snps)) 
raw_SZ #16 individuals and 227501 SNPs. 
filtered_SZ #16 individuals and 223601 SNPs.
r2vcftools::Chrom(snps) #return chromosome (contig), position, and ID of each SNP.
```

### 1.3 Remove individuals IF IT IS NECESSARY

In my case, let's assume I don't remove any individuals from the vcf file.
```{r}
# If you do not need to remove individuals,
snps_unind = snps
VCFsummary(snps_unind) ## 16 individuals and 223601 SNPs.
```

### 1.4 verify the quality of data

If the saved snps_unind object is successful, I can resume anytime from here.
```{r eval=FALSE, include=FALSE}
#access the output from working directory
#saveRDS(snps_unind, "./output/snps_unind.rds")
#snps_unind <- readRDS("./output/snps_unind.rds")
```

```{r}
#See Genotype Matrix
genotypes <- r2vcftools::GenotypeMatrix(snps_unind)
genotypes[1:10, 1:10] ## -1 is missing; otherwise, gives the number of derived alleles in the genotype -- ie. a 0 and 2 are both homozygotes 
#saveRDS(genotypes, "./output/genotypes.rds")
#genotypes <- readRDS("./output/genotypes.rds")
```

```{r}
#Look at depth, quality, HWE, HE, allele frequencies, and Pi

site.depth <- r2vcftools::Query(snps_unind, type="site-mean-depth")
summary(site.depth$MEAN_DEPTH) #Mean = 17 / Median = 12
hist(site.depth$MEAN_DEPTH, breaks=200) 

quality <- Query(snps_unind, type="site-quality") 
summary(quality$QUAL) #all 13 with seemingly no variation
hist(quality$QUAL) ## something wrong here.
#---------------------------------- analysed till here Nov 17th, 2020
PI <- Query(snps_unind, type="site-pi")
mean(PI$PI) ## Mean nucleotide divergency per-site ###0.29
hist(PI$PI)

HWE <- Query(snps_unind, type="hardy")
summary(HWE$P_HWE) #Mean = 0.129 / Median = 0.0012
hist(HWE$P_HWE)
hist(HWE$P_HWE[HWE$P_HWE<0.0001])

HE <- Query(snps_unind, type="het")
hist(HE$O.HOM) ## O.HOM, E.HOM, N_SITES, F
hist(HE$E.HOM) 
hist(HE$N_SITES) 
hist(HE$F) ## Inbreeding coefficient 0.4, quite high

freq <- Query(snps_unind, type="freq2")
hist(freq$X.FREQ.)
hist(freq$X.FREQ.1)
head(freq)
```

```{r eval=FALSE, include=FALSE}
#missing per locus
Missing <- apply(GenotypeMatrix(snps_unind), 2, function(x) sum(x < 0)/length(x)*100)
summary(Missing) ## no missingness??
hist(Missing)
```

```{r eval=FALSE, include=FALSE}
#missing per individual
Missing_ind <- apply(GenotypeMatrix(snps_unind),1, function(x) sum(x<0)/length(x)*100)
summary(Missing_ind) ## Max missing = 0
hist(Missing_ind)
```

### 1.5 add geographic info by individuals to a vcf file

```{r}
#H. Add coords and missing information to vcf file. Be careful to the order!

#add the information. You can add more columns if you need, like ecological traits or other grouping classification 
snps_unind@meta$local_ID = coords[,4] #locality pond ID
snps_unind@meta$longitude = coords[,3] #longitude
snps_unind@meta$latitude = coords[,2] #latitude
snps_unind@meta$ind_ID = coords[,1] #sample ID
#verify the file
head(snps_unind@meta) #verify
tail(snps_unind@meta) #verify
```
```{r}
snps_ind_pc <- snps_unind
```




### TreeMix